bigquery/snippets/table/bigquery_inserting_data_types.go

```diff
@@ -0,0 +1,106 @@
+
+package table
+
+// [START bigquery_inserting_data_types]
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"cloud.google.com/go/civil"
+)
+
+// ComplexType represents a complex row item
+type ComplexType struct {
+	Name         string                 `bigquery:"name"`
+	Age          int                    `bigquery:"age"`
+	School       []byte                 `bigquery:"school"`
+	Location     bigquery.NullGeography `bigquery:"location"`
+	Measurements []float64              `bigquery:"measurements"`
+	DatesTime    DatesTime              `bigquery:"datesTime"`
+}
+
+// DatesTime shows different date/time representation
+type DatesTime struct {
+	Day        civil.Date     `bigquery:"day"`
+	FirstTime  civil.DateTime `bigquery:"firstTime"`
+	SecondTime civil.Time     `bigquery:"secondTime"`
+	ThirdTime  time.Time      `bigquery:"thirdTime"`
+}
+
+// insertingDataTypes demonstrates inserting data into a table using the streaming insert mechanism.
+func insertingDataTypes(projectID, datasetID, tableID string) error {
+	// projectID := "my-project-id"
+	// datasetID := "mydataset"
+	// tableID := "mytable"
+	ctx := context.Background()
+	client, err := bigquery.NewClient(ctx, projectID)
+	if err != nil {
+		return fmt.Errorf("bigquery.NewClient: %w", err)
+	}
+	defer client.Close()
+
+	// Manually defining schema
+	schema := bigquery.Schema{
+		{Name: "name", Type: bigquery.StringFieldType},
+		{Name: "age", Type: bigquery.IntegerFieldType},
+		{Name: "school", Type: bigquery.BytesFieldType},
+		{Name: "location", Type: bigquery.GeographyFieldType},
+		{Name: "measurements", Type: bigquery.FloatFieldType, Repeated: true},
+		{Name: "datesTime", Type: bigquery.RecordFieldType, Schema: bigquery.Schema{
+			{Name: "day", Type: bigquery.DateFieldType},
+			{Name: "firstTime", Type: bigquery.DateTimeFieldType},
+			{Name: "secondTime", Type: bigquery.TimeFieldType},
+			{Name: "thirdTime", Type: bigquery.TimestampFieldType},
+		}},
+	}
+	// Infer schema from struct
+	// schema, err := bigquery.InferSchema(ComplexType{})
+
+	table := client.Dataset(datasetID).Table(tableID)
+	err = table.Create(ctx, &bigquery.TableMetadata{
+		Schema: schema,
+	})
+	if err != nil {
+		return fmt.Errorf("table.Create: %w", err)
+	}
+	day, _ := civil.ParseDate("2019-01-12")
+	firstTime, _ := civil.ParseDateTime("2019-02-17T11:24:00.000")
+	secondTime, _ := civil.ParseTime("14:00:00")
+	thirdTime, _ := time.Parse(time.RFC3339Nano, "2020-04-27T18:07:25.356Z")
+	row := &ComplexType{
+		Name:         "Tom",
+		Age:          30,
+		School:       []byte("Test University"),
+		Location:     bigquery.NullGeography{GeographyVal: "POINT(1 2)", Valid: true},
+		Measurements: []float64{50.05, 100.5},
+		DatesTime: DatesTime{
+			Day:        day,
+			FirstTime:  firstTime,
+			SecondTime: secondTime,
+			ThirdTime:  thirdTime,
+		},
+	}
+	rows := []*ComplexType{row}
+	inserter := table.Inserter()
+	if err := inserter.Put(ctx, rows); err != nil {
+		return err
```

Comment:
It may be worth showing how to check for a [MultiError](https://pkg.go.dev/cloud.google.com/go/bigquery#MultiError) here, for users who are generating a mix of good and bad rows.

Reply:
I came up with two versions for showing error handling but no sure which one to use:

1. Try to insert a list of rows, with one being a bad one, prints the errors, but the function itself return error as nil because we are skipping invalid rows.
```
rows := []*ComplexType{row}
badRow := &ComplexType{
    Name: "John",
    Age:  24,
    // this row is missing required fields
}
rows = append(rows, badRow)

inserter := table.Inserter()
inserter.SkipInvalidRows = true
err = inserter.Put(ctx, rows)
if err != nil {
    if multiErr, ok := err.(bigquery.PutMultiError); ok {
        for _, putErr := range multiErr {
            fmt.Printf("failed to insert row %d with err: %v \n", putErr.RowIndex, putErr.Error())
        }
        return nil
    }
    return err
}
return nil
```

2.  Add a commented `badRow`, keep the default behavior of not skipping invalid rows and return an error if the MultiError is found:

```
rows := []*ComplexType{row}
/* Uncomment to simulate insert errors
badRow := &ComplexType{
    Name: "John",
    Age:  24,
    // this row is missing required fields
}
rows = append(rows, badRow)*/

inserter := table.Inserter()
err = inserter.Put(ctx, rows)
if err != nil {
    if multiErr, ok := err.(bigquery.PutMultiError); ok {
        for _, putErr := range multiErr {
            fmt.Printf("failed to insert row %d with err: %v \n", putErr.RowIndex, putErr.Error())
        }
    }
    return err
}
return nil
```

Reply:
Both approaches look good to me. I have a slight preference for the second one, just because the sample works out of the box, and lets a user "break it" if they want to.

Also, nit: if you do go with the second approach use `//` style comments which are somewhat easier to remove in my opinion.

Reply:
cool, I went with the second approach and just push a change to the comment style